File: sliding window/76. Minimum Window Substring.js
`Given two strings s and t of lengths m and n respectively,
 return the minimum window substring of s such that every character in t (including duplicates) is included in the window. 
 If there is no such substring, return the empty string "".

The testcases will be generated such that the answer is unique.

A substring is a contiguous sequence of characters within the string.

 

Example 1:

Input: s = "ADOBECODEBANC", t = "ABC"
Output: "BANC"
Explanation: The minimum window substring "BANC" includes 'A', 'B', and 'C' from string t.
Example 2:

Input: s = "a", t = "a"
Output: "a"
Explanation: The entire string s is the minimum window.
Example 3:

Input: s = "a", t = "aa"
Output: ""
Explanation: Both 'a's from t must be included in the window.
Since the largest window of s only has one 'a', return empty string.
 

Constraints:

m == s.length
n == t.length
1 <= m, n <= 105
s and t consist of uppercase and lowercase English letters.
 

Follow up: Could you find an algorithm that runs in O(m + n) time?`;

const minWindow = (s, t) => {
  // ! if t length is greater than s length, return empty string
  if (s.length < t.length) return "";
  const t_map = new Map();
  const window_map = new Map();

  for (const char of t) {
    t_map.set(char, t_map.get(char) + 1 || 1);
  }

  let have = 0;
  let need = t_map.size;

  let resLen = Number.MAX_VALUE;
  let res = [-1, -1];

  let window_start = 0;

  for (let i = 0; i < s.length; i++) {
    const char = s[i];

    if (t_map.has(char)) {
      window_map.set(char, window_map.get(char) + 1 || 1);
      if (window_map.get(char) === t_map.get(char)) {
        have++;
      }
    }
    // ! if have is equal to need, we have a valid window
    while (have === need) {
      //! time to update the result
      if (i - window_start + 1 < resLen) {
        res = [window_start, i];
        resLen = i - window_start + 1;
      }
      //! now we are going to pop the char from left
      const leftChar = s[window_start];
      if (t_map.has(leftChar)) {
        window_map.set(leftChar, window_map.get(leftChar) - 1);
        if (window_map.get(leftChar) < t_map.get(leftChar)) {
          have--;
        }
      }
      window_start++;
    }
  }
  //! if no valid window, return empty string else return the substring
  return resLen === Number.MAX_VALUE ? "" : s.slice(res[0], res[1] + 1);
};

console.log(minWindow("ADOBECODEBANC", "ABC"));

File: sliding window/76. Minimum Window Substring.py

from collections import Counter
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        # base cases
        if not s or not t:
            return ""
        if len(s) < len(t):
            return ""
        
        # create a dictionary of the characters in t
        t_dict = Counter(t)

        window= {}

        have, need = 0, len(t_dict)

        res,resLen=[-1,-1], float('inf') # [start, end]

        left=0

        for right in range(len(s)):
            curr=s[right]
            window[curr]=window.get(curr,0)+1

            if curr in t_dict and window[curr]==t_dict[curr]:
                have+=1

            while have==need:
                curr_size=right-left+1
                # we found the smallest window
                if curr_size<resLen:
                    res=[left,right]
                    resLen=curr_size
                # move the left pointer
                window[s[left]]-=1
                # if the character is in the dictionary and the window has less than the required amount
                if s[left] in t_dict and window[s[left]]<t_dict[s[left]]:
                    have-=1

                left+=1

        if res[0]==-1:
            return ""
        
        return s[res[0]:res[1]+1]
    
# Time complexity: O(n)
# Space complexity: O(1) since the size of the window is fixed and the size of the dictionary is fixed
    


# Test cases
test = Solution()
print(test.minWindow("ADOBECODEBANC", "ABC") )
